File: output/playwright/nested-wires/capture-m3.py
"""Reproduce Chromium evidence through the installed Playwright CLI skill. Never starts Vite."""
import json
import subprocess
from pathlib import Path

ROOT = Path(__file__).resolve().parents[3]
EVIDENCE = ROOT / 'output/playwright/nested-wires'
CLI = Path.home() / '.codex/skills/playwright/scripts/playwright_cli.sh'


def command(*args):
    result = subprocess.run([str(CLI), '-s=nested-wires', *args], cwd=ROOT,
                            text=True, capture_output=True, check=True)
    if '### Error' in result.stdout:
        raise RuntimeError(result.stdout)
    return result.stdout


command('open', 'http://127.0.0.1:5188/roads-prototype.html?nested', '--headed')
command('snapshot')
source = (EVIDENCE / 'capture-m3.mjs').read_text().replace('export async function', 'async function')
output = command('run-code', 'async () => {\n' + source + '\nreturn capture(page);\n}')
report = json.loads(output.split('### Result\n', 1)[1].split('\n###', 1)[0])
(EVIDENCE / 'm3-browser.json').write_text(json.dumps(report, indent=2) + '\n')
print(f"PASS M3 five loads={[l['roads:navigation-to-ready'] for l in report['loads']]}; median={report['medianMilliseconds']:.3f} ms; M2=243.7 ms; ceiling=280 ms")
assert report['medianMilliseconds'] <= 280, report
assert report['wires'] == 18 and report['nodes'] == 22 and report['layoutRecalcCount'] == 1
print('PASS 22 nodes, 18 wires, one layout calculation; four M3 screenshots captured')

File: output/playwright/nested-wires/capture-m3.mjs
/** Browser evidence on the existing Vite. Export through capture-m3.py; failures reach Node. */
const directory = 'output/playwright/nested-wires/';
async function shot(page, name, focus) {
  await page.getByRole('combobox', { name: 'Wire focus' }).selectOption(focus);
  await page.waitForTimeout(200);
  await page.screenshot({ path: `${directory}${name}.png` });
  return { name, focus, viewport: page.viewportSize() };
}
export async function capture(page) {
  await page.setViewportSize({ width: 1920, height: 1440 });
  const loads = [];
  for (let i = 0; i < 5; i += 1) {
    await page.goto('http://127.0.0.1:5188/roads-prototype.html?nested');
    await page.waitForFunction(
      () => performance.getEntriesByName('roads:navigation-to-ready').length > 0,
    );
    loads.push(
      await page.evaluate(() =>
        Object.fromEntries(
          performance
            .getEntriesByType('measure')
            .filter((e) => e.name.startsWith('roads:'))
            .map((e) => [e.name, e.duration]),
        ),
      ),
    );
  }
  const screenshots = [];
  screenshots.push(await shot(page, 'm3-overview', ''));
  screenshots.push(await shot(page, 'm3-corridor-s2s4', 'w10'));
  screenshots.push(await shot(page, 'm3-corridor-s3s4', 'w17'));
  await page.setViewportSize({ width: 2560, height: 1920 });
  await page.waitForTimeout(250);
  screenshots.push(await shot(page, 'm3-shared-rows', 'w12'));
  return {
    loads,
    medianMilliseconds: loads
      .map((l) => l['roads:navigation-to-ready'])
      .toSorted((a, b) => a - b)[2],
    screenshots,
    wires: await page.locator('[data-wire-id]').count(),
    nodes: await page.locator('[data-node-id]').count(),
    layoutRecalcCount: await page.evaluate(() => window.__layoutRecalcCount),
  };
}
